#
# This patch make sure we load all lazy loaded libraries from plugins;
# these are not loaded for plugins.
#
--- old/session.c	2023-04-06 16:50:47.339502741 +0200
+++ new/session.c	2023-04-06 16:51:29.661284370 +0200
@@ -1352,6 +1352,61 @@ do_nologin(struct passwd *pw)
 	exit(254);
 }
 
+#ifdef LOAD_ALL_LIBS
+/*
+ * Check which plugins we have loaded; reload them again bit this time we
+ * load them with RTLD_NOW so we have everything loaded on chroot() and
+ * ld.so will no longer bail out.
+ * Libraries loaded with sshd directly are always loaded using "-z now".
+ */
+
+#include <dlfcn.h>
+#include <dirent.h>
+
+
+static void
+load_all_libs(void)
+{
+	DIR *d = opendir("/proc/self/path");
+	struct dirent *ent;
+
+	if (d == NULL) {
+		debug("cannot open /proc/self/path");
+		return;
+	}
+
+	while ((ent = readdir(d)) != NULL) {
+		char p[MAXPATHLEN];
+		ssize_t s;
+
+		/*
+		 * Ignore fds, cwd, root, ".", ".." and files in tmpfs.
+		 */
+		if (ent->d_name[0] == '.' || isdigit(ent->d_name[0]) ||
+		    strcmp(ent->d_name, "a.out") == 0 ||
+		    strcmp(ent->d_name, "cwd") == 0 ||
+		    strcmp(ent->d_name, "root") == 0 ||
+		    strncmp(ent->d_name, "tmpfs.", 6) == 0) {
+			continue;
+		}
+		if ((s = readlinkat(dirfd(d), ent->d_name, p, sizeof (p))) < 0) {
+			continue;
+		}
+		p[s] = '\0';
+		/* nsswitch and pam modules */
+		if (strstr(p, "/nss_") != NULL || strstr(p, "/security/") != NULL) {
+			void *lib;
+			debug3("preloading %s", p);
+			lib = dlopen(p, RTLD_NOW|RTLD_NODELETE);
+			if (lib != NULL) {
+				(void) dlclose(lib);
+			}
+		}
+	}
+	(void) closedir(d);
+}
+#endif
+
 /*
  * Chroot into a directory after checking it for safety: all path components
  * must be root-owned directories with strict permissions.
@@ -1396,6 +1451,10 @@ safely_chroot(const char *path, uid_t ui
 
 	}
 
+#ifdef LOAD_ALL_LIBS
+	load_all_libs();
+#endif
+
 	if (chdir(path) == -1)
 		fatal("Unable to chdir to chroot path \"%s\": "
 		    "%s", path, strerror(errno));
